File: java/src/ch1/ex117.java
package ch1;

public class ex117 {
    public static void main(String[] args) {
        Times func = (int i, int i2) -> mult_invariant(i, i2);

        System.out.println(func.apply(9, 5));
        System.out.println(func.apply(9, 8));
        System.out.println(func.apply(9, 12));
        System.out.println(func.apply(9, 99));
        System.out.println("done");
    }

    @FunctionalInterface
    interface Times {
        int apply(int i, int i2);
    }

    static int mult_invariant(int a, int b) {
        if (b == 0) return 0;
        if (even(b)) return 2 * mult_invariant(a, b / 2);
        return a + mult_invariant(a, b - 1);
    }

    static boolean even(int x) {
        return x % 2 == 0;
    }
}

File: ruby/ch1/ex117.rb
module Ex117

def even(x)
  x % 2 == 0
end

def mult_invariant(a, b)
  if b==0
    0
  elsif even(b)
    2 * mult_invariant(a, b/2)
  else
    a + mult_invariant(a, b-1)
  end
end

def mult(a, b)
  mult_invariant(a, b)
end

end #Ex117

include Ex117
print "Testing\n"
print mult(5, 4) , "\n"
print mult(5, 5) , "\n"
print mult(5, 10) , "\n"
print mult(5, 20) , "\n"




  
